SecuritySystemUWP/SecuritySystemUWP/Storage/OneDrive.cs

```diff
@@ -158,5 +154,98 @@ private void SaveTokens(object sender, string arg)
+            bool success = await parseResponse(response, tryNumber);
+            var events = new Dictionary<string, string>();
+
+            if (success)
+            {
+                numberUploaded++;
+                await file.DeleteAsync();
+                this.lastUploadTime = DateTime.Now;
+            }
+            else if (tryNumber <= MaxTries)
+            {
+                events.Add("Retrying upload", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToUploadPicture", events);
+                await uploadWithRetry(file, ++tryNumber);
+            }
+            else
+            {
+                events.Add("Max upload attempts reached", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToUploadPicture - total failure", events);
+            }
+        }
+
+        private async Task deleteWithRetry(string picture, string folder, int tryNumber)
+        {
+            HttpResponseMessage response = await oneDriveConnector.DeleteFileAsync(picture, folder);
+            bool success = await parseResponse(response, tryNumber);
+            var events = new Dictionary<string, string>();
+
+            if (success)
+            {
+                //no additional action needed
+            }
+            else if (tryNumber <= MaxTries)
+            {
+                events.Add("Retrying delete", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToDeletePicture", events);
+                await deleteWithRetry(picture, folder, ++tryNumber);
+            }
+            else
+            {
+                events.Add("Max delete attempts reached", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToDeletePicture - total failure", events);
+            }
+        }
+
+        private async Task<bool> parseResponse(HttpResponseMessage response, int tryNumber)
+        {
+            bool successStatus;
+            var events = new Dictionary<string, string>();
+            events.Add("Attempt number", tryNumber.ToString());
+
+            if (response.IsSuccessStatusCode)
+            {
+                successStatus = true;
+                TelemetryHelper.TrackEvent("HttpSuccess", events);
+            }
+            else
+            {
+                successStatus = false;
+                events.Add(response.StatusCode.ToString(), response.ReasonPhrase);
+                events.Add("Original Request", response.RequestMessage.ToString());
+                if (response.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    events.Add("Recovery Attempt", "AuthorizeWithRefreshToken()");
+                    if (App.Controller.XmlSettings.OneDriveRefreshToken == "")
+                    {
+                        events.Add("App.Controller.XmlSettings.OneDriveRefreshToken", "Invalid Empty String");

```

Comment:
Should we also add a oneDriveConnector.Reauthorize() call hear to attempt to recover from this empty string state?